# A Notebook to Use Naïve Bayes Classifiers

This notebook shows how to train a Naïve Bayes classifier to classify unseen instances.
 
* The variable "dataset" stores the name of text file that you input and is passed as an argument of the function "loadDataSet()".  
* After processing, the loadDataSet function will output, or in other words, return two variables, "instances", and "labels".  
* "instances" stores the feature vector of each instance. "labels" stores the labels of all instances.   
* The variable "n_foldCV" stores the number of times of n-fold cross validation that you input.
* The variable "choice" stores your choice of classifiers. "instances", "labels" and "choice" are the arguments of function chooseClassifier and the function return a variable "clf".
* The variable "clf" stores up to three Naive Bayes models, and it can be fitted with "instances" and "labels". Once the model are trained, they can be used to predict unseen instances. 

In [1]:
import numpy as np
from sklearn.naive_bayes import BernoulliNB,GaussianNB,MultinomialNB
from sklearn.model_selection import cross_val_score

def loadDataSet(dataset): 
    with open(dataset) as f:
        data=f.readlines()
        attributes=data[0].rstrip().split(',')[:-1]
        instances=[entry.rstrip().split(',')[:-1] for entry in data[1:]]
        dataArray=[]
        Dict={}
        for i in range(len(instances[0])):
            try:
                dataArray.append([float(instance[i]) for instance in instances])
            except:
                encodedData,vocab=encode([instance[i] for instance in instances])
                dataArray.append(encodedData)
                Dict[i]=vocab
                print(attributes[i],': ',list(vocab.items()))
        instances=np.array(dataArray).T
        labels=[entry.rstrip().split(',')[-1] for entry in data[1:]]
        return instances,labels,Dict

def encode(data):
    vocab={}
    uniqueVals=list(set(data))
    for Val in uniqueVals:
        vocab[Val]=uniqueVals.index(Val)
    encodedData=list(map(uniqueVals.index,data))
    return encodedData,vocab


def predict(testset):
    if "clf_G" in globals():
        prediction=clf_G.predict(testset)
        print("GaussianNB: ",prediction)
    if "clf_M" in globals():
        prediction=clf_M.predict(testset)
        print("MultinomialNB: ",prediction)

## Training: Build a Naïve Bayes Classifier##
The cell below asks for a dataset. It trains a Naïve Bayes classifier. There are three Naive Bayes classifiers provided. They are based on different math fundations and might have different performance over different datasets.   

We provide two classification datasets that could be applied to the Naïve Bayes algorithms. 
* ["iris.txt"](https://archive.ics.uci.edu/ml/datasets/iris) has four attributes with continuous values describing three different iris species.
* ["lenses.txt"](https://archive.ics.uci.edu/ml/datasets/lenses) contains four attributes with discrete values and three classes.

### Gaussian Naïve Bayes Classifier

In [2]:
dataset=input("Please Enter Your Data Set:")
instances,labels,Dict=loadDataSet(dataset)
clf_G = GaussianNB()
clf_G.fit(instances, labels)
print("Gaussian Naive Bayes is used.")

Please Enter Your Data Set:C:/Users/Sam/Desktop/INF549/iris.txt
Gaussian Naive Bayes is used.


### Multinomial Naive Bayes Classifier

In [3]:
dataset=input("Please Enter Your Data Set:")
instances,labels,Dict=loadDataSet(dataset)
clf_M = MultinomialNB()
clf_M.fit(instances, labels)
print("Multinomial Naive Bayes is used.")

Please Enter Your Data Set:C:/Users/Sam/Desktop/INF549/iris.txt
Multinomial Naive Bayes is used.


## Predict unseen instances##
When you are prompted to input a test set, please create an example of an instance that looks like the instances in the training set.  

Each feature value is separated with a comma, and should have the same length as the instances in the training set. 

In [5]:
testset=input('Please Enter Your Test Set:')
testset=testset.split(",")
temp=[]
for i in range(len(testset)):
    try:
        temp.append(float(testset[i]))
    except:
        temp.append(Dict[i][testset[i]])
testset=np.array(temp).reshape((1,len(temp)))

Please Enter Your Test Set:4.8,3,1.1,0.2


In [6]:
predict(testset)

GaussianNB:  ['Iris-setosa']
MultinomialNB:  ['Iris-setosa']


## Evaluate a classifier##
The following cell will output the accuracy score in each run and the accuracy estimate of the model under 95% confidence interval.  

### Gaussian Naïve Bayes Classifier

In [7]:
dataset=input('Please Enter Your Test Data:')
n_foldCV=int(input("Please Enter the Number of Folds:"))
instances,labels,Dict=loadDataSet(dataset)
clf_G = GaussianNB()
scores = cross_val_score(clf_G, instances, labels, cv=n_foldCV)
print("======GaussianNB======")
print(np.array2string(scores,separator=","))
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Please Enter Your Test Data:C:/Users/Sam/Desktop/INF549/iris.txt
Please Enter the Number of Folds:10
======GaussianNB======
[0.93333333,0.93333333,1.        ,0.93333333,0.93333333,0.93333333,
 0.86666667,1.        ,1.        ,1.        ]
Accuracy: 0.95 (+/- 0.09)


### Multinomial Naive Bayes

In [8]:
dataset=input('Please Enter Your Test Data:')
n_foldCV=int(input("Please Enter the Number of Folds:"))
instances,labels,Dict=loadDataSet(dataset)
clf_M = MultinomialNB()
scores = cross_val_score(clf_M, instances, labels, cv=n_foldCV)
print("======MultinomialNB======")
print(np.array2string(scores,separator=","))
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Please Enter Your Test Data:C:/Users/Sam/Desktop/INF549/iris.txt
Please Enter the Number of Folds:10
======MultinomialNB======
[1.        ,1.        ,1.        ,0.93333333,0.86666667,0.93333333,
 0.8       ,1.        ,1.        ,1.        ]
Accuracy: 0.95 (+/- 0.13)


### Extra Credit:  Bernoulii Naïve Bayes Classifier
Try to train a Bernoulli Naïve Bayes classifier on any of the dataset and report its performance. Does it perform well or not? If not, can you explain why it doesn't work well on the dataset and give some solutions? You don't need to write codes to implement your solutions, just write them down in your report.

In [14]:
dataset=input('Please Enter Your Test Data:')
n_foldCV=int(input("Please Enter the Number of Folds:"))
instances,labels,Dict=loadDataSet(dataset)
clf_B = BernoulliNB()
scores = cross_val_score(clf_B, instances, labels, cv=n_foldCV)
print("======BernoulliNB======")
print(np.array2string(scores,separator=","))
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Please Enter Your Test Data:C:/Users/Sam/Desktop/INF549/Homework 3/Datasets/iris.txt
Please Enter the Number of Folds:10
======BernoulliNB======
[0.33333333,0.33333333,0.33333333,0.33333333,0.33333333,0.33333333,
 0.33333333,0.33333333,0.33333333,0.33333333]
Accuracy: 0.33 (+/- 0.00)




Guassian Naive Bayes model assumes the data to be Guassian or Normal distribution with mean=0 SD=1. It is generally used for continous data.
Multiomial Naive Bayes model counts the frequency of the events that occur.
Bernoulli Naive Bayes model is similar to Multinomial with the difference being instead of counting the frequency of the event occurred, it only describes whether or not an event occurred.
In the example above I ran Iris Data set in both Q1 and Q2 to compare the accuracies if various classifiers. Iris dataset cointains real valued data hence Gaussian would be more suitable. 
Given these characterstics of the model we can infer that the Bernoulli Naive Bayes model does not seem to be the correct choice for a dataset like Iris.

Bernoulli is used in cases with binary features. Multinom treats features as event probablities. Since iris is non binary and real valued Guassian and multinom works best for the dataset.

When we use lenses.txt the accuracy improves to 72% from 33%(for Iris). Since Bernoulli is designed for binary/boolean features. If it encounters any other kind of data a BernoulliNB will binarize its input i.e., there may be multiple features but each one is assumed to be a binary-valued (Bernoulli, boolean) variable. Lenses is categorical whereas Iris is real valued hence the accuracy is better with respect to lenses data set
